### PR TITLE
Darreldonald devops

### PR DESCRIPTION
**Pull request by DarrelDonald**
Added a client and controller to manage DevOps repositories for the detector graduation project
Has functions to retrieve code, view branches, make pull requests, and push changes

### CODE REVIEW

File: src/Diagnostics.RuntimeHost/Services/DevOpsClient/DevOpsClient.cs
Comment (rekhaswaminathan) [inline src/Diagnostics.RuntimeHost/Services/DevOpsClient/DevOpsClient.cs:49]: Is `_config` used anywhere else? I see that you are passing constructor parameter `config` to `LoadConfigurations` method

File: src/Diagnostics.RuntimeHost/Services/DevOpsClient/DevOpsClient.cs
Comment (rekhaswaminathan) [inline src/Diagnostics.RuntimeHost/Services/DevOpsClient/DevOpsClient.cs:52]: `dev.azure.com` may not be the prefix every time. In some cases it changes based on a team's set up. Eg; it can be msazure.visualstudio.com/<org>/<repo>. I recommend storing the entire domain as config and pull from that. In the future we can save each partner's URL as a config property.

**Review by rekhaswaminathan — CHANGES_REQUESTED** [review]

File: src/Diagnostics.RuntimeHost/Controllers/DevOpsController.cs
Comment (rekhaswaminathan) [inline src/Diagnostics.RuntimeHost/Controllers/DevOpsController.cs:33]: How does AppLens client know which target branch to make PR against?

File: src/Diagnostics.RuntimeHost/Services/DevOpsClient/DevOpsClient.cs
Comment (DarrelDonald) [reply src/Diagnostics.RuntimeHost/Services/DevOpsClient/DevOpsClient.cs:49]: It does not. I meant to remove it.

File: src/Diagnostics.RuntimeHost/Controllers/DevOpsController.cs
Comment (DarrelDonald) [reply src/Diagnostics.RuntimeHost/Controllers/DevOpsController.cs:33]: It gets it from user input

File: src/Diagnostics.RuntimeHost/Services/DevOpsClient/DevOpsClient.cs
Comment (DarrelDonald) [reply src/Diagnostics.RuntimeHost/Services/DevOpsClient/DevOpsClient.cs:52]: I tried accessing a repo on `msazure.visualstudio.com` with the API using `dev.azure.com` and was successful. I believe this format will work even if the domain is different.

Comment (rekhaswaminathan) [reply src/Diagnostics.RuntimeHost/Services/DevOpsClient/DevOpsClient.cs:50]: Thats great. Then we can stick to this

File: src/Diagnostics.RuntimeHost/Controllers/DevOpsController.cs
Comment (rekhaswaminathan) [inline src/Diagnostics.RuntimeHost/Controllers/DevOpsController.cs:26]: Are you using the `configuration` parameter anywhere?

File: src/Diagnostics.RuntimeHost/Services/DevOpsClient/DevOpsClient.cs
Comment (rekhaswaminathan) [inline src/Diagnostics.RuntimeHost/Services/DevOpsClient/DevOpsClient.cs:36]: I see that `IRepoClient` is the one with the methods defined. Shouldn't that be the interface here?

File: src/Diagnostics.RuntimeHost/Services/DevOpsClient/DevOpsClient.cs
Comment (rekhaswaminathan) [inline src/Diagnostics.RuntimeHost/Services/DevOpsClient/DevOpsClient.cs:23]: What is the purpose of this interface?

File: src/Diagnostics.RuntimeHost/Startup.cs
Comment (rekhaswaminathan) [inline src/Diagnostics.RuntimeHost/Startup.cs:164]: Shouldn't it be `IRepoClient` here instead of `IDevOpsClient`? 
`IDevOpsClient` seems to have only some properties defined but no methods.

**Review by rekhaswaminathan — CHANGES_REQUESTED** [review]

**Comment by rekhaswaminathan** [comment]
For all the APIs in the new Controller, consider passing the resource URI or at least the Resource Provider & Resource Type as post body object or part of query param because in the future not every resource will end up having same Devops Repo. We have to switch between the Repos based on the service caller or the resource type.

File: src/Diagnostics.RuntimeHost/Controllers/DevOpsController.cs
Comment (ShekharGupta1988) [inline src/Diagnostics.RuntimeHost/Controllers/DevOpsController.cs:32]: can resource Uri also be moved to json body ?

File: src/Diagnostics.RuntimeHost/Controllers/DevOpsController.cs
Comment (ShekharGupta1988) [inline src/Diagnostics.RuntimeHost/Controllers/DevOpsController.cs:36]: This parameter validation in json body is a repeated pattern I see in multiple controllers.
Can you create a utility method to do this.
Something like :

bool ValidateRequestBody(JToken jsonBody, string[] fieldNames, out string outputMessage = null);

File: src/Diagnostics.RuntimeHost/Services/DevOpsClient/DevOpsClient.cs
Comment (rekhaswaminathan) [reply src/Diagnostics.RuntimeHost/Services/DevOpsClient/DevOpsClient.cs:105]: Hey @DarrelDonald , any reason for sending back gitClient HttpStatusCode back to the client? The client would ideally need to parse out only the result object, correct? Is there a need for Runtimehost to simply echo back the `gitClient.LastResponseContext.HttpStatusCode`?

File: src/Diagnostics.RuntimeHost/Services/DevOpsClient/DevOpsClient.cs
Comment (rekhaswaminathan) [inline src/Diagnostics.RuntimeHost/Services/DevOpsClient/DevOpsClient.cs:159]: Do we need to throw a `new VssServiceException()` here? Can we just rethrow the original exception? This may end up changing the stack trace when its captured by RuntimeHost's default unhandled exception event logging.

File: src/Diagnostics.Logger/ETWProvider/DiagnosticsETWProvider.cs
Comment (rekhaswaminathan) [inline src/Diagnostics.Logger/ETWProvider/DiagnosticsETWProvider.cs:285]: I believe this would require the Event attribute & its properties to be given. You can refer to other logging methods to see how its constructed.

**Review by rekhaswaminathan — APPROVED** [review]

**Comment by rekhaswaminathan** [comment]
Hey @DarrelDonald, changes look good! I'd get @ShekharGupta1988's approval as well before merging

**Review by ShekharGupta1988 — APPROVED** [review]

**Comment by ShekharGupta1988** [comment]
Looks good to me as well. Just shared some comments offline with Darrel regarding the ETW event parameters. Rest all is good :)

**Pull request merged**
